### PR TITLE
HDDS-1597. Remove hdds-server-scm dependency from ozone-common

### PR DESCRIPTION
**Pull request by elek**
I noticed that the hadoop-ozone/common project depends on hadoop-hdds-server-scm project.

The common projects are designed to be a shared artifacts between client and server side. Adding additional dependency to the common pom means that the dependency will be available for all the clients as well.

(See the attached artifact about the current, desired structure).

We definitely don't need scm server dependency on the client side.

The code dependency is just one class (ScmUtils) and the shared code can be easily moved to the common.

See: https://issues.apache.org/jira/browse/HDDS-1597

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 44 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 67 | Maven dependency ordering for branch |
| +1 | mvninstall | 530 | trunk passed |
| +1 | compile | 286 | trunk passed |
| +1 | checkstyle | 87 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 888 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 136 | trunk passed |
| 0 | spotbugs | 285 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 483 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 35 | Maven dependency ordering for patch |
| +1 | mvninstall | 457 | the patch passed |
| +1 | compile | 267 | the patch passed |
| +1 | javac | 267 | the patch passed |
| -0 | checkstyle | 35 | hadoop-hdds: The patch generated 5 new + 0 unchanged - 0 fixed = 5 total (was 0) |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | xml | 3 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 620 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 147 | the patch passed |
| +1 | findbugs | 501 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 144 | hadoop-hdds in the patch failed. |
| -1 | unit | 1055 | hadoop-ozone in the patch failed. |
| -1 | asflicense | 46 | The patch generated 17 ASF License warnings. |
| | | 6044 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.container.common.statemachine.commandhandler.TestBlockDeletion |
|   | hadoop.ozone.container.common.impl.TestContainerPersistence |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineProvider |
|   | hadoop.ozone.client.rpc.TestFailureHandlingByClient |
|   | hadoop.ozone.client.rpc.TestBCSID |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-860/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/860 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle xml |
| uname | Linux 7f4b33b13374 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 4a692bc |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/1/artifact/out/diff-checkstyle-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/1/testReport/ |
| asflicense | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/1/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 4585 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-hdds/framework hadoop-hdds/server-scm hadoop-ozone/common hadoop-ozone/integration-test hadoop-ozone/ozone-manager hadoop-ozone/tools U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java [hunk 1]
```diff
@@ -35,28 +35,27 @@
...
-import org.apache.hadoop.hdds.conf.OzoneConfiguration;
-import org.apache.hadoop.hdds.protocol.SCMSecurityProtocol;
 import org.apache.hadoop.hdds.scm.protocolPB.ScmBlockLocationProtocolPB;
 import org.apache.hadoop.ipc.Client;
 import org.apache.hadoop.ipc.ProtobufRpcEngine;
 import org.apache.hadoop.ipc.RPC;
 import org.apache.hadoop.metrics2.util.MBeans;
 import org.apache.hadoop.net.DNS;
 import org.apache.hadoop.net.NetUtils;
+import org.apache.hadoop.security.UserGroupInformation;
```
Comment (bharatviswa504) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java]: I see no changes in this file, just re-ordering of imports.
Why is this change needed here?

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/ServerUtils.java [hunk 2]
```diff
@@ -203,4 +203,16 @@ public static void setOzoneMetaDirPath(OzoneConfiguration conf,
     conf.set(HddsConfigKeys.OZONE_METADATA_DIRS, path);
   }
 
+  public static File getDBPath(Configuration conf, String dbDirectory) {
+    final File dbDirPath =
+        getDirectoryFromConfig(conf, dbDirectory, "OM");
+    if (dbDirPath != null) {
+      return dbDirPath;
```
Comment (bharatviswa504) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/ServerUtils.java]: Add java doc for this method.

File: hadoop-ozone/integration-test/pom.xml [hunk 3]
```diff
@@ -60,6 +64,11 @@ https://maven.apache.org/xsd/maven-4.0.0.xsd">
       <groupId>org.apache.hadoop</groupId>
       <artifactId>hadoop-ozone-client</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.apache.commons</groupId>
+      <artifactId>commons-lang3</artifactId>
```
Comment (bharatviswa504) [inline hadoop-ozone/integration-test/pom.xml]: Why is this additional dependency brought in?

Comment (bharatviswa504) [reply hadoop-ozone/integration-test/pom.xml]: Understood, this is because of the below test needs some classes from this jar.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestKeyManagerImpl.java [hunk 4]
```diff
@@ -52,6 +51,7 @@
 import org.apache.hadoop.test.GenericTestUtils;
 import org.apache.hadoop.test.LambdaTestUtils;
 
+import org.apache.commons.lang3.RandomStringUtils;
```
Comment (bharatviswa504) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestKeyManagerImpl.java]: Same in this, I don't see any code changes only re-ordering of imports,
I think we don't need this change in this PR.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/ServerUtils.java (hunk 2 above)
Comment (elek) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/ServerUtils.java]: Thanks. I added it.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java (hunk 1 above)
Comment (elek) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java]: Ups, you are right. I reverted it.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestKeyManagerImpl.java (hunk 4 above)
Comment (elek) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestKeyManagerImpl.java]: In this case this is not just a reorder because I replaced the original commons-lang2 transitive dependency with commons-lang3 and the package name is different.

But I agree, this is not a fair modification. Let's fix only the dependencies here and don't include additional modification. I fixed it with using the original commons-lang2 (and added the explicit dependency which was used from a transitive dependency until now).

**Comment by elek** [comment]
Thanks the review @bharatviswa504. I improved the patch as you suggested (see the separated commits)

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestKeyManagerImpl.java (hunk 4 above)
Comment (bharatviswa504) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestKeyManagerImpl.java]: Ya you are right, I missed that..

**Comment by elek** [comment]
> I have a question, why we explicilty added commons-lang as dependency with scope test here? Which we have not done before. Is this for a reason so that down stream who depend on server-scm don't get this dependency. Because previously without this patch, we don't have anywhere commons-lang defined in our pom.xml. I think it is transitive dependency from one of the jars.

Very good question. I just realized that it can't be compiled and I added the dependency without checking the reason ;-)

I tried to answer your question and analyzed the dependencies but it turned out that it's not about the transitive dependencies:

I moved the TestKeymanagerImpl from ozone-manager project to the integration-test project because it requires the the scm classes for testing. Until now we collected all of these unit tests (which require multiple daemon projects and starts real servers like the MiniOzoneCluster) in the integration-test project, so I think it's the best place anyway (and it was required because I removed the scm dependency of om).

And common-lang2 was a dependency of OM but not the integration-test.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 29 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 59 | Maven dependency ordering for branch |
| -1 | mvninstall | 255 | hadoop-ozone in trunk failed. |
| -1 | compile | 129 | hadoop-ozone in trunk failed. |
| +1 | checkstyle | 71 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 817 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 161 | trunk passed |
| 0 | spotbugs | 193 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 224 | hadoop-ozone in trunk failed. |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 28 | Maven dependency ordering for patch |
| -1 | mvninstall | 197 | hadoop-ozone in the patch failed. |
| -1 | compile | 137 | hadoop-ozone in the patch failed. |
| -1 | javac | 137 | hadoop-ozone in the patch failed. |
| +1 | checkstyle | 93 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | xml | 5 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 664 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 139 | the patch passed |
| -1 | findbugs | 230 | hadoop-ozone in the patch failed. |
||| _ Other Tests _ |
| +1 | unit | 227 | hadoop-hdds in the patch passed. |
| -1 | unit | 194 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 50 | The patch does not generate ASF License warnings. |
| | | 4666 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-860/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/860 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle xml |
| uname | Linux f1906e6ec9c0 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 18c1eeb |
| Default Java | 1.8.0_212 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/4/artifact/out/branch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/4/artifact/out/branch-compile-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/4/artifact/out/branch-findbugs-hadoop-ozone.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/4/artifact/out/patch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/4/artifact/out/patch-compile-hadoop-ozone.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/4/artifact/out/patch-compile-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/4/artifact/out/patch-findbugs-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/4/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/4/testReport/ |
| Max. process+thread count | 1327 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-hdds/framework hadoop-hdds/server-scm hadoop-ozone hadoop-ozone/common hadoop-ozone/integration-test hadoop-ozone/ozone-manager hadoop-ozone/tools U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-860/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by bharatviswa504** [comment]
> > I have a question, why we explicilty added commons-lang as dependency with scope test here? Which we have not done before. Is this for a reason so that down stream who depend on server-scm don't get this dependency. Because previously without this patch, we don't have anywhere commons-lang defined in our pom.xml. I think it is transitive dependency from one of the jars.
> 
> Very good question. I just realized that it can't be compiled and I added the dependency without checking the reason ;-)
> 
> I tried to answer your question and analyzed the dependencies but it turned out that it's not about the transitive dependencies:
> 
> I moved the TestKeymanagerImpl from ozone-manager project to the integration-test project because it requires the the scm classes for testing. Until now we collected all of these unit tests (which require multiple daemon projects and starts real servers like the MiniOzoneCluster) in the integration-test project, so I think it's the best place anyway (and it was required because I removed the scm dependency of om).
> 
> And common-lang2 was a dependency of OM but not the integration-test.

Thank You @elek for the explanation. I think previously for OM it might be coming through a transitive dependency. As no where in the pom.xml commons-lang was explicitly defined.

**Comment by elek** [comment]
Thanks @bharatviswa504 the review and commit. I noticed a problem with ratis/THREE replication. I can't write any key and no error is visible on the console log.

I tried to revert multiple commits one by one and this one is the most suspicious one. I will revert it for now, and retest it in a separated PR to be sure it's fine.
